pages/docs/reference/scope-functions.md

```diff
@@ -0,0 +1,576 @@
+```kotlin
+fun processNonNullString(str: String) {}
+
+fun main() {
+//sampleStart
+    val str: String? = "Hello"   
+    //processNonNullString(str)       // compilation error: str can be null
+    val length = str?.let { 
+        println("let() called on $it")        
+        processNonNullString(it)      // OK: the function is called only on non-nulls
+        it.length
+    }
+//sampleEnd
+}
+```
+</div>
+
+Another case for using `let` is introducing local variables with a limited scope for improving code readability. To define a new variable for the context object, provide its name as the lambda argument so that it can be used instead of the default `it`.
+
+<div class="sample" markdown="1" theme="idea">
+```kotlin
+fun main() {
+//sampleStart
+    val numbers = listOf("one", "two", "three", "four")
+    val modifiedFirstItem = numbers.first().let { firstItem ->
+        println("The first item of the list is '$firstItem'")
+        if (firstItem.length >= 5) firstItem else "!" + firstItem + "!"
+    }.toUpperCase()
+    println("First item after modifications: '$modifiedFirstItem'")
+//sampleEnd
+}
+```
+</div>
+
+### `with`
+
+A non-extension function: **the context object** is passed as an argument, but inside the lambda, it's available as a receiver (`this`). **The return value** is the lambda result. 
+
+We recommend `with` for calling functions on the context object without providing the lambda result. In the code, `with` can be read as “_with this object, do the following._”
+
+<div class="sample" markdown="1" theme="idea">
+```kotlin
+fun main() {
+//sampleStart
+    val numbers = mutableListOf("one", "two", "three")
+    with(numbers) {
+        println("'with' is called with argument $this")
+        println("It contains $size elements")
+    }
+//sampleEnd
+}
+```
+</div>
+
+Another use case for `with` is introducing a helper object whose properties or functions will be used for calculating a value.
+
+<div class="sample" markdown="1" theme="idea">
+```kotlin
+fun main() {
+//sampleStart
+    val numbers = mutableListOf("one", "two", "three")
+    val firstAndLast = with(numbers) {
+        "The first element is ${first()}," +
+        " the last element is ${last()}"
+    }
+    println(firstAndLast)
+//sampleEnd
+}
+```
+</div>
+
+### `run`
+
+**The context object** is available as a receiver (`this`). **The return value** is the lambda result.
+
+`run` does the same as `with` but invokes as `let` - as an extension function of the context object.
+
+`run` is useful when your lambda contains both the object initialization and the computation of the return value.
+
+<div class="sample" markdown="1" theme="idea">
+```kotlin
+data class Person(var name: String, var age: Int = 0, var city: String = "")
+
+fun main() {
+//sampleStart
+    val isTeenager = Person("Adam").run {
+        age = 32
+        city = "London"
+        age > 12 && age < 20
+    }
+    println("Is Adam a teenager? $isTeenager")
```

Comment:
I can't help but feel that this `run` sample is quite contrived compared to its original use case from [KT-5235](https://youtrack.jetbrains.com/issue/KT-5235). I'm tempted to rewrite it with `apply` for setting the properties + `run` for evaluating the expression, but `apply` has not been introduced at this moment.

Perhaps this one could illustrate it better:
```
    class MultiportService(var url: String, var port: Int) {
        fun prepareRequest(): String = "Default request"
        fun query(request: String): String = "Result for query '$request'"
    }
    
// start
    val service = MultiportService("https://example.kotlinlang.org", 80)
      
    val result = service.run {
        port = 8080
        query(prepareRequest() + ", and some extra")
    }
    // an alternative with 'let' could be shown
// end
    println(result)
```
The idea here is the necessity to call several methods of a context object that depend on each other's result or the state of object they mutate.

Reply:
Thanks, I've replaced the example with this (with small modifications and the alternative with `let`).